### PR TITLE
Changed the float validation inside DrawInfo.ApplyTransform for performance gain

### PR DESCRIPTION
**Pull request by FreezyLemon**
Saw #1153 and tried to reduce CPU usage of the function. In my (isolated) tests with a huge load of tested floats, this resulted in around 12-15% performance increase and in in-game tests (granted I'm not too used to performance profiling, but the numbers should still be correct) CPU time for checkComponentValid was actually reduced by 15-20%.

Explanation for why I did what I did:
The old implementation checks for `float.IsNaN(value) || float.IsInfinity(value)`. So for this to be true, the `value` has to be `float.NaN`, `float.PositiveInfinity` or `float.NegativeInfinity`.

Looking into the specifications, if the exponent of a float value is 255 (its max value), the float has to be any one of these three values ([source](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#Exponent_encoding), last row). So in this new implementation I'm checking whether the exponent is 255 and if it is, the float is invalid by the old standards.

As to what the code does:
Everything except the `singleToExponentAsByte` method should be self-explanatory.

`*(int*)(&value)` takes `value`'s pointer, casts it to an int pointer and dereferences the pointer so we get an int back. This int has the same binary contents as `value`.

This is done so that we can shift the exponent bits (bits 2 - 8) to the 8 least significant bits (`>> 23`), which allows us to cast this int to a byte (so a smaller data type is passed around) which contains the exponent of `value`.

### CODE REVIEW

**Comment by Tom94** [comment]
Hey there, nice idea! Based on your description I'm not sure whether there was an _overall_ improvement of 15% or just a 15% speed-up of the `checkComponentValid` call. I think the more important question to answer is how your change compares to leaving out `checkComponentValid` completely, i.e. what is the _slowdown_ compared to not doing the check at all. Also, just to clarify, this should be measured in release mode.

File: osu.Framework/Graphics/DrawInfo.cs [hunk 1]
```diff
@@ -82,18 +82,25 @@ public void ApplyTransform(Vector2 translation, Vector2 scale, float rotation, V
...
-            if (float.IsNaN(component.X) || float.IsInfinity(component.X))
+            if (isInvalidFloat(component.X))
                 throw new ArgumentException($"Invalid value ({component}) provided for component {name}.X.");
-            if (float.IsNaN(component.Y) || float.IsInfinity(component.Y))
+            if (isInvalidFloat(component.Y))
                 throw new ArgumentException($"Invalid value ({component}) provided for component {name}.Y.");
         }
 
+        // Takes the bits from value, interprets them as an int and then shifts them so that we get the exponent (bit 2 to 8) back.
+        // Returns as byte because it's a smaller data type and enough for the exponent
```
Comment (Tom94) [inline osu.Framework/Graphics/DrawInfo.cs]: This comment could be XML-doc.

File: osu.Framework/Graphics/DrawInfo.cs [hunk 2]
```diff
@@ -82,18 +82,25 @@ public void ApplyTransform(Vector2 translation, Vector2 scale, float rotation, V
...
+        private unsafe byte singleToExponentAsByte(float value) => (byte)(*(int*)&value >> 23);
+
+        // If exponent is 255, float is either ±Infinity or NaN (by definition)
```
Comment (Tom94) [inline osu.Framework/Graphics/DrawInfo.cs]: Likewise this one.

File: osu.Framework/Graphics/DrawInfo.cs [hunk 3]
```diff
@@ -82,18 +82,25 @@ public void ApplyTransform(Vector2 translation, Vector2 scale, float rotation, V
...
+        private bool isInvalidFloat(float toCheck) => singleToExponentAsByte(toCheck) == byte.MaxValue;
```
Comment (Tom94) [inline osu.Framework/Graphics/DrawInfo.cs]: How about naming this method `isFinite`? NaN and Inf _are_ technically valid floats, and `isfinite` is also used for this functionality in the [C++ standard library](http://en.cppreference.com/w/cpp/numeric/math/isfinite).

**Comment by FreezyLemon** [comment]
Also, to clarify the initial statement:
The new `isFinite` (or checkValidComponent) is about 15% faster than the old `checkComponentValid` implementation in an isolated testing environment (literally just [in]validating 500 million random floats with both methods and taking the time). In the actual game (yes, in release mode. yes, also without debugger attached, just the performance profiler), the performance gain seemed to be a bit more for some reason...

Haven't done tests without any kind of checks yet.

**Comment by peppy** [comment]
to be honest i'm more interested in gains under debug (especially in mono, the overhead of debug is huge at the moment). that sounds pretty worthwhile, though.

File: osu.Framework/Graphics/Drawable.cs [hunk 4]
```diff
@@ -925,6 +932,8 @@ public virtual Vector2 OriginPosition
 
             set
             {
+                if (!isFinite(value)) throw new ArgumentException($@"{nameof(OriginPosition)} must be finite, but is {value}");
```
Comment (Tom94) [inline osu.Framework/Graphics/Drawable.cs]: You probably also want to add such a check to `RelativeAnchorPosition`. This is what I meant with that it becomes harder to cover all code paths. :P

Comment (FreezyLemon) [reply osu.Framework/Graphics/Drawable.cs]: I don't see how that would be necessary if we're just trying to cover all previous paths.. How does `RelativeAnchorPosition` intersect with any of the `ApplyTransform` parameters? I really don't see it

Comment (Tom94) [reply osu.Framework/Graphics/Drawable.cs]: `AnchorPosition` (which relates multiplicatively to `RelativeAnchorPosition`) is an additive component of the position input passed to `ApplyTransform`. Pretty short computation path actually.

Comment (FreezyLemon) [reply osu.Framework/Graphics/Drawable.cs]: Oh, now I see it... how could I have missed that! I'll blame it on being tired...

**Comment by peppy** [comment]
Good job on these changes 👍

**Pull request merged**